### ISSUE TITLE
Device Authorization Grant

### ISSUE THREAD

**Issue by oliamb** [django-oauth/django-oauth-toolkit#962]
Hey, thanks a lot for the awesome library! 

I just discovered this evening an extension to OAuth2, which is the Device Authorization Grant, defined in RFC https://tools.ietf.org/html/rfc8628.

I dig a bit on how rfc8628 relates to rfc6749 (OAuth2). I thought it was through the [Native Application](https://tools.ietf.org/html/rfc8252) update at first. But no, they are just leveraging the [extension grant](https://tools.ietf.org/html/rfc6749#section-4.5) from OAuth2, which is not supported either by the _django-oauth-toolkit_.

The Device Authorization grant looks super interesting to authenticate native applications, much simpler than Authorization Code with PKCE and inter-application URL redirections. I am wondering if supporting it directly or through supporting the extension grant has been discussed already? (I did not find any reference)

**Comment by dopry** [comment]
We would be happy accept a PR for this feature once it is available in OAuthLib.  We'll need to wait on upstream to merge it. It looks like there is ongoing work in https://github.com/oauthlib/oauthlib/pull/844 and there is upstream interest. If you're interested in this feature I would recommend supporting the work in OAuthLib first.

**Comment by tykling** [comment]
Great to see that it got merged upstream! :+1: :rocket:

**Comment by dopry** [comment]
@oliamb, et al. With this merged upstream we would love to see a PR for this feature. I don't have time to work on the implementation, but I would make time to support anyone who wanted to undertake the endeavor.

**Comment by duzumaki** [comment]
@dopry I'll be raising the pr in DOT after the oauthlib stuff is released

**Comment by duzumaki** [comment]
@dopry I'll be raising the pr in DOT after the oauthlib stuff is released. I've just raised another pr in oauthlib to add the necessary device code grant type object and the relevant methods.

We implement our own auth server where I work, so I have a working version using the branches I made in DOT and oauthlib

**Comment by duzumaki** [comment]
@dopry I'll be raising the pr in DOT after the oauthlib stuff is released. I've just raised another pr in oauthlib to add the necessary device code grant type object and the relevant methods.

I have a working device flow end to end using the branches I made in DOT and oauthlib with a custom auth server I'm working on

**Comment by dopry** [comment]
Feel free to open a draft PR before the upstream release. That would allow @n2ygk and I to do some initial code review.

**Comment by duzumaki** [comment]
draft [here](https://github.com/jazzband/django-oauth-toolkit/pull/1539) @dopry
